EcoSoapBank/Model Controller/PickupController.swift

```diff
@@ -9,14 +9,51 @@
 import Foundation
 
 
+protocol PickupDataProvider {
+    func fetchAllPickups(_ completion: (Result<[Pickup], Error>) -> Void)
+}
+
+
 class PickupController: ObservableObject {
-    @Published var pickups: [Pickup] = .mockData()
+    @Published private(set) var pickups: [Pickup] = []
+    @Published private(set) var error: Error?
+
+    private var dataProvider: PickupDataProvider
+
+    init(dataProvider: PickupDataProvider) {
+        self.dataProvider = dataProvider
+
+        self.dataProvider.fetchAllPickups { [weak self] result in
+            switch result {
+            case .success(let pickups):
+                self?.pickups = pickups
+            case .failure(let error):
+                self?.error = error
+            }
+        }
+    }
 }
 
 
-extension Array where Element == Pickup {
-    static func mockData() -> Self {
-        [
+// MARK: - Mock Data Provider
+
+struct MockPickupProvider: PickupDataProvider {
+    enum Error: Swift.Error {
+        case shouldFail
+    }
+
+    var shouldFail: Bool
```

Comment:
I gather this is for simulating a network failure when using the mock data?

Reply:
Yeah, exactly. I suppose that should use a comment lol.